raiden/network/proxies/user_deposit.py

```diff
@@ -577,3 +658,161 @@ def _deposit_check_result(
+                    f"Couldn't plan withdraw because planned amount "
+                    f"{amount_to_plan_withdraw} exceeded current balance of {current_balance}."
+                )
+
+            raise RaidenRecoverableError("Plan withdraw failed for an unknown reason.")
+
+        return transaction_mined
+
+    def _withdraw_check_result(
+        self,
+        transaction_sent: Optional[TransactionSent],
+        amount_to_withdraw: TokenAmount,
+        token: Token,
+        previous_token_balance: TokenAmount,
+    ) -> None:
+        if transaction_sent is None:
+            failed_at = self.client.get_block(BLOCK_ID_LATEST)
+            failed_at_blocknumber = failed_at["number"]
+
+            self.client.check_for_insufficient_eth(
+                transaction_name="withdraw",
+                transaction_executed=False,
+                required_gas=self.gas_measurements["UserDeposit.withdraw"],
+                block_identifier=failed_at_blocknumber,
+            )
+            raise RaidenRecoverableError(
+                "Withdraw transaction failed to be sent for an unknown reason."
+            )
+
+        transaction_mined = self.client.poll_transaction(transaction_sent)
+
+        if not was_transaction_successfully_mined(transaction_mined):
+            failed_at_blocknumber = BlockNumber(transaction_mined.receipt["blockNumber"])
+
+            withdraw_plan = self.get_withdraw_plan(
+                withdrawer_address=self.node_address, block_identifier=failed_at_blocknumber
+            )
+            whole_balance = self.whole_balance(block_identifier=failed_at_blocknumber)
+
+            if amount_to_withdraw > withdraw_plan.withdraw_amount:
+                raise RaidenRecoverableError(
+                    f"Couldn't withdraw {amount_to_withdraw}, "
+                    f"current withdraw plan only allows for {withdraw_plan.withdraw_amount}."
+                )
+
+            if withdraw_plan.withdraw_block > failed_at_blocknumber:
+                raise RaidenRecoverableError(
+                    f"Couldn't withdraw at block {failed_at_blocknumber}. "
+                    f"The current withdraw plan requires block number "
+                    f"{withdraw_plan.withdraw_block}."
+                )
+
+            if whole_balance - amount_to_withdraw < 0:
+                raise RaidenRecoverableError(
+                    f"The current whole balance is {whole_balance}. "
+                    f"The withdraw of {amount_to_withdraw} would have lead to an underflow."
+                )
+
+            current_token_balance = TokenAmount(
+                token.balance_of(self.node_address, block_identifier=failed_at_blocknumber)
+            )
+            # FIXME: This seems fishy. The token balance could have an unexpected value for
+            #        unrelated reasons (e.g. concurrent token transfers via transferFrom).
+            if current_token_balance != previous_token_balance + amount_to_withdraw:
```

Comment:
As the comment says this seems slightly prone to false positives. The same is true for other places in here.

Reply:
note that the `UserDeposit` *has* require messages, we don't have that on some of the smart contracts because of bytecode limits, but here the error message can just be inferred from that.

Reply:
Unfortunately that doesn't work correctly in web3py when using proxies. Karl is working on it though: https://github.com/ethereum/web3.py/issues/941